### PR TITLE
Wrote enough tests to cover the models file (now at 84% coverage over…

### LINKED ISSUE #268: Writing tests
- Reach 80-90% test coverage
- Make sure all devs know how to write tests (host a workshop?)
- New requirements for devs writing Features: must have full test coverage on new code
  - Could this be verified in the build?

This issue can be broken down into several if necessary.

### PR DESCRIPTION
**Pull request by borjainglefernandez**
## Status
- Done

## GitHub Issues addressed (`#IssueNumber`)
- This PR closes #268 

## What I did
- Wrote enough tests to completely cover models.py
- Modified existing tests

## Screenshots
- Before
- After

## Tests
- A brief explanation of tests done/written or how reviewers can test your work

## Questions/Discussions/Notes
- Eventually need to write more tests
- PR's that modify models (will have to update tests to reflect those changes)

## Merging the PR
- Who should merge this PR?
  - [ ] I will merge it myself
  - [ ] The last reviewer to approve can merge it
- Should the commit history be preserved in the base branch, or is it okay to combine all of them into a single commit ("squash-merge")?
  - [ ] preserve the history
  - [ ] squash-merge

### CODE REVIEW

File: tcf_website/tests/test_instructor_edgecases.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            self.instructor.average_group_hours_for_course(
+                self.course) is None)
+
+    def test_average_other_hours_for_course_no_reviews(self):
+        """Test average other hours for a particular course when there are no reviews."""
+
+        self.review1.delete()
+        self.review2.delete()
+
+        self.assertTrue(
```
Comment (borjainglefernandez) [inline tcf_website/tests/test_instructor_edgecases.py:111]: I was thinking we should change all of these assertTrue's to assertEqual's since if the test fails assertTrue does not lend any information as to what the two sides of the equality are whereas assertEqual does.

File: tcf_website/tests/test_review.py [hunk 2]
```diff
@@ -126,3 +129,84 @@ def test_unauthorized_user_delete(self):
...
+
+    def test_upvote(self):
+        """Test for upvote method verify with count_votes"""
+
+        self.review1.upvote(self.user4)
+        expected_sum = 2
+        self.assertEqual(self.review1.count_votes(), expected_sum)
+
+    def test_upvote_already_upvoted(self):
+        """Test for upvote method verify with count_votes"""
```
Comment (Bartleby2718) [inline tcf_website/tests/test_review.py:160]: This docstring is the same as the one for the method above.

File: tcf_website/tests/test_review.py [hunk 3]
```diff
@@ -126,3 +129,84 @@ def test_unauthorized_user_delete(self):
...
+    def test_double_vote(self):
+        """Test for voting twice on same review by same user using vote model"""
+        try:
+            self.assertRaises(IntegrityError, Vote.objects.create(
+                value=-1,
+                user=self.user3,
+                review=self.review1))
+
+        except IntegrityError:
+            pass
```
Comment (Bartleby2718) [inline tcf_website/tests/test_review.py:191]: If you have this try/except block because of the error message, you can use `suppress_request_warnings` instead.

File: tcf_website/tests/test_subdepartment.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        self.assertFalse(self.subdepartment.has_current_course())
+
+
+def create_new_sem(obj, year):
+    """Helper method to modify current semester"""
+    obj.semester = Semester.objects.create(
+        year=year,
+        season='FALL',
+        number=1508
+    )
```
Comment (Bartleby2718) [inline tcf_website/tests/test_subdepartment.py:52]: I believe this helper method should be either in `SubdepartmentTestCase` or in `test_utils.py`.

File: tcf_website/tests/test_user.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# pylint: disable=no-member
+# pylint: disable=missing-module-docstring
```
Comment (Bartleby2718) [inline tcf_website/tests/test_user.py:2]: Can we have docstring instead? I believe the number of exceptions should be minimal.

File: tcf_website/tests/test_utils.py [hunk 6]
```diff
@@ -30,38 +30,42 @@ def setup(obj):
         season='FALL',
         number=1108
     )
+    obj.incomplete_semester = Semester.objects.create(
+        year=2019,
+        number=1198)
```
Comment (Bartleby2718) [inline tcf_website/tests/test_utils.py:35]: I think this should be either
```suggestion
    obj.incomplete_semester = Semester.objects.create(year=2019, number=1198)
```
or
```suggestion
    obj.incomplete_semester = Semester.objects.create(
        year=2019,
        number=1198,
    )
```

File: tcf_website/tests/test_utils.py [hunk 7]
```diff
@@ -70,16 +74,76 @@ def setup(obj):
         username="tcf2yay",
         email="tcf2yay@virginia.edu",
         computing_id="tcf2yay",
+        first_name="Taylor",
+        last_name="Comb",
```
Comment (Bartleby2718) [inline tcf_website/tests/test_utils.py:78]: If you think about it, a person named `Taylor Comb` cannot have the computing ID `tcf2yay`...

**Review by Bartleby2718 — CHANGES_REQUESTED** [review]
Great job! It must have been very tedious writing these tests. I wrote a few comments that can be applied to other pieces of code, so if you do decide to make any changes, make sure to do the same for the rest.

File: tcf_website/tests/test_utils.py (hunk 7 above)
Comment (j-alicia-long) [reply tcf_website/tests/test_utils.py:78]: ^I know this isn't srs but just FYI my official computing ID is rz5sc and my legal name is Jennifer Long

so don't put in a spam check for mismatched computing IDs pls

File: tcf_website/tests/test_course.py [hunk 8]
```diff
@@ -15,10 +15,16 @@ class CourseTestCase(TestCase):
     def setUp(self):
         setup(self)
 
+    def test_review_count(self):
+        """Test review count method"""
+
+        expected_reviews = 2
```
[the remark this answers is not in the document]
Comment (borjainglefernandez) [reply tcf_website/tests/test_course.py:21]: Yeah I think I'll just put the 2 in the assertEquals

File: tcf_website/tests/test_course.py [hunk 9]
```diff
@@ -15,10 +15,16 @@ class CourseTestCase(TestCase):
...
+        self.assertEqual(self.course.review_count(), expected_reviews)
+
     def test_code(self):
         """Course code string."""
         code = self.course.code()
-        self.assertTrue(code == 'CS 420')
+        self.assertTrue(code == 'CS 1420')
```
[the remark this answers is not in the document]
Comment (borjainglefernandez) [reply tcf_website/tests/test_course.py:27]: Yes, this was not a test I wrote I simply just changed the code of the class to ensure it passes. I'll look through all the test files and change the one's that should be assertTrue to assertEquals.

File: tcf_website/tests/test_instructor.py [hunk 10]
```diff
@@ -40,21 +40,143 @@ def test_average_difficulty(self):
...
+            self.instructor.average_gpa_for_course(
+                self.course), average_gpa_course)
+
+    def test_taught_courses(self):
+        """Test taught courses for a particular section."""
+
+        courses_taught = "<QuerySet [<Section: CS 1420 | Software Testing | Fall 2020 | " \
+                         "Tom Jefferson (tjt3rea@virginia.edu)>, <Section: CS 1421 | Algorithms " \
+                         "| Fall 2020 | Tom Jefferson (tjt3rea@virginia.edu)>]>"
+        self.assertEqual(str(self.instructor.taught_courses()), courses_taught)
```
[the remark this answers is not in the document]
Comment (borjainglefernandez) [reply tcf_website/tests/test_instructor.py:139]: The short answer is no. I have not yet found a way to mock a queryset object for testing (Django should really fix this) but I'll keep looking.

File: tcf_website/tests/test_instructor.py [hunk 11]
```diff
@@ -40,21 +40,143 @@ def test_average_difficulty(self):
...
+            / 6
+
+        avg_enjoyability = (self.review1.enjoyability + self.review2.enjoyability +
+                            self.review3.enjoyability + self.review4.enjoyability +
+                            self.review5.enjoyability + self.review6.enjoyability) / 6
+
+        avg_rating = (avg_instructor_rating +
+                      avg_recommendability + avg_enjoyability) / 3
+
+        self.assertEqual(self.instructor.average_rating(), avg_rating)
```
[the remark this answers is not in the document]
Comment (borjainglefernandez) [reply tcf_website/tests/test_instructor.py:159]: Yes, I should be using assertAlmostEqual instead.

File: tcf_website/tests/test_instructor_edgecases.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    """Tests for instructor model with edge cases."""
+
+    def setUp(self):
+        setup(self)
+
+    def test_average_rating_for_course_no_reviews(self):
+        """Test average rating for a particular course when there are no reviews."""
+
+        self.review1.delete()
+        self.review2.delete()
```
[the remark this answers is not in the document]
Comment (borjainglefernandez) [reply tcf_website/tests/test_instructor_edgecases.py:19]: The thing is there are other reviews created used for other tests and although that one-liner would not interrupt anything, it might be better practice to show the specific reviews that are part of that course.

File: tcf_website/tests/test_instructor_edgecases.py [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+        """Test average difficulty for all instructor's courses when there are no reviews."""
+
+        self.assertTrue(self.instructor2.average_difficulty() is None)
+
+    def test_taught_courses_none(self):
+        """Test taught courses when there are none."""
+
+        courses_taught = "<QuerySet []>"
+        self.assertEqual(
+            str(self.instructor2.taught_courses()), courses_taught)
```
[the remark this answers is not in the document]
Comment (borjainglefernandez) [reply tcf_website/tests/test_instructor_edgecases.py:135]: Yes

File: tcf_website/tests/test_review.py (hunk 2 above)
Comment (borjainglefernandez) [reply tcf_website/tests/test_review.py:160]: Good catch! As you can tell most of the time I was copy and pasting and changing them to fit the current test cause I got lazy

File: tcf_website/tests/test_review.py (hunk 3 above)
Comment (borjainglefernandez) [reply tcf_website/tests/test_review.py:191]: I tried that and it does not work. I think it is because it isn't a request error it's an actual exception. But now I figured out how to do it without the try except block.

**Comment by borjainglefernandez** [comment]
I think I fixed all of the issues @Bartleby2718 pointed out. Thank you so much for being thorough, it really helped create the best possible tests. Some highlights: I figured out how to model query sets so that we are not comparing the strings, I changed all the assertTrue that should have been assertEquals, I changed some assertEqual to assertAlmostEqual with 4 decimal place precision.

File: tcf_website/tests/test_utils.py [hunk 14]
```diff
@@ -70,16 +75,76 @@ def setup(obj):
...
+
+    obj.section_course = Section.objects.create(
+        course=obj.course,
+        semester=obj.semester,
+        sis_section_number=312312
+    )
+
+    obj.section_course.instructors.set(
+        Instructor.objects.filter(
+            first_name="Tom"))
```
Comment (Bartleby2718) [inline tcf_website/tests/test_utils.py:112]: Did you mean to use a `QuerySet` that includes the `Tom Jefferson` instance above (i.e. `obj.instructor`) or any instructors with the first name `Tom`? If the former is the case, the following code seems more appropriate, since someone might add another `Instructor` with `first_name` `Tom`:
```suggestion
    obj.section_course.instructors.set(
        Instructor.objects.filter(
            pk=obj.instructor.pk))
```
If you do decide to change this, make sure to change this elsewhere.

File: tcf_website/tests/test_utils.py [hunk 15]
```diff
@@ -96,7 +161,7 @@ def setup(obj):
         amount_group=20,
         amount_reading=20,
         amount_writing=20,
-        amount_homework=20
+        amount_homework=20,
```
Comment (Bartleby2718) [inline tcf_website/tests/test_utils.py:164]: I don't usually tell people to use or not use trailing commas (although I always use them when a list of items spans multiple lines), but this seems inconsistent given that you didn't use trailing commas elsewhere (e.g. `obj.instructor_grade`). I think this diff is unnecessary, unless you want to enforce a consistent coding style for _all_ such cases.
```suggestion
        amount_homework=20
```
I personally believe minimizing diffs helps future developers maintain the same code base because it reduces the number of diff they have to check. If there are too many unnecessary diffs in a commit/PR (although that's not the case in this PR), one is less likely to look closely into them.
If you do commit this suggestion, make sure to change `obj.review2` as well.

File: tcf_website/tests/test_utils.py [hunk 16]
```diff
@@ -202,3 +290,12 @@ def new_function(*args, **kwargs):
...
 
     return new_function
+
+
+def create_new_sem(obj, year):
+    """Helper method to modify current semester"""
+    obj.semester = Semester.objects.create(
+        year=year,
+        season='FALL',
+        number=1508
```
Comment (Bartleby2718) [inline tcf_website/tests/test_utils.py:300]: Just an FYI: this `number` is a concatenation of `1`, the last two digits of `year`, and the starting month of the `season`. This _might_ confuse developers if there is a test fixture that does not conform to this. Should be something along this line:
```suggestion
        number=f'1{year % 100}8'

File: tcf_website/tests/test_utils.py [hunk 17]
```diff
@@ -202,3 +290,12 @@ def new_function(*args, **kwargs):
         logger.setLevel(previous_logging_level)
 
     return new_function
+
+
+def create_new_sem(obj, year):
```
Comment (Bartleby2718) [inline tcf_website/tests/test_utils.py:295]: I looked through all the diffs in this PR, but this seems like the only case where an abbreviation is used. I don't think adding `ester` at the end will hurt.

**Review by Bartleby2718 — CHANGES_REQUESTED** [review]
Seems like I missed quite a lot in the first review. I'm sorry to have you fix this twice, but I hope this review is helpful since I spent a full hour on this.

File: tcf_website/tests/test_subdepartment.py [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+
+        self.assertQuerysetEqual(
+            self.subdepartment.recent_courses(),
+            recent_courses,
+            transform=lambda x: x,  # Needed so that the formatting works
+            ordered=False)
+
+    def test_recent_courses_has_no_recent_courses(self):
+        """Test recent_courses method in Subdepartment model when there are not recent courses"""
+        create_new_sem(self, 2050)
```
[the remark this answers is not in the document]
Comment (borjainglefernandez) [reply tcf_website/tests/test_subdepartment.py:40]: That does not work as the create_new_sem (now create_new_semester) function is only in test_utils.py as a static function.

**Review by Bartleby2718 — APPROVED** [review]
Great work!
